Support Buffers and fuel mode for WASM UDFs (#105574)
ClickHouse/ClickHouse: docs/en/sql-reference/functions/wasm_udf.md

- `DETERMINISTIC`: Declares the function as deterministic — always returns the same output for the same input. When specified, ClickHouse may constant-fold calls where all arguments are constants: the function is evaluated once at query analysis time and the result is reused for every row.
- `SHA256_HASH`: Expected module hash for verification (auto-filled if omitted), can be used to ensure the correct WASM module loaded across different replicas.
- `SETTINGS`: Per-function settings

>     - `serialization_format` String — Serialization format for ABI requires it. Supported values: `MsgPack`, `JSONEachRow`, `CSV`, `TSV`, `TSVRaw`, `RowBinary`, and `Buffers`. Default: `MsgPack`. Under `ABI BUFFERED_V1`, `Buffers` uses the declared function schema and ClickHouse Native per-column binary encoding. When `serialization_format = 'Buffers'`, the guest must return a single-column `Buffers` block whose `num_rows` equals the input block row count and whose column type matches the declared `RETURNS` type.

Reviewer: ```suggestion
    - `serialization_format` String — Serialization format for ABI requires it. Supported values: `MsgPack`, `JSONEachRow`, `CSV`, `TSV`, `TSVRaw`, `RowBinary`, and `Buffers`. Default: `MsgPack`. Block-based formats such as `Buffers` must return a single column whose type match the declared function signature.
```

>     - `webassembly_udf_enable_fuel` Bool — Enables finite fuel budgeting for the function. Default: `true`. Set at function creation time via `CREATE FUNCTION ... LANGUAGE WASM ... SETTINGS ...`; the value is persisted with the function definition. When `false`, the query-level setting `webassembly_udf_max_fuel` is ignored for this function. `webassembly_udf_max_memory` still applies. Query cancellation via `max_execution_time`, `KILL QUERY`, and shutdown still applies. For `webassembly_udf_engine = 'wasmedge'`, ClickHouse keeps WasmEdge cost measuring active internally to deliver cancellation; for Wasmtime, epoch interruption remains enabled.

Reviewer: Looks a bit wordy and includes extra implementation details, imo

```suggestion
    - `webassembly_udf_enable_fuel` Bool — Enables finite fuel budgeting for the function. Default: `true`. When `false`, the query-level setting `webassembly_udf_max_fuel` is ignored for this function. Disabling fuel limits may improve performance when using the `wasmtime` engine. However, for untrusted or buggy guest code, it can increase the risk of runaway execution.
```

Reviewer: Also, what is the main motivation behind introducing this setting at the function level?

Possible alternatives I considered

* Server level - Pros: simple configuration model. Cons: no per-module differentiation, and enabling it may require changing server config.

* Module level - Allows distinguishing between “safe” and “less trusted” modules. The downside is that configuration may become awkward, since modules are added via `INSERT INTO ...`; this would likely require attaching settings metadata via Map column for example, which is somewhat unconventional.

* Function level - current approach

* Runtime-only decision based on `webassembly_udf_max_fuel = 0` - Very flexible, but would require always storing two compiled versions of each module, increasing memory usage even for cases when we use only single fuel mode.

And another dimension here is access control - who is allowed to disable fuel limits: the server administrator / the module / the function creator or even the query user (this could also potentially be restricted through existing settings constraints mechanisms).

Curious to hear your thoughts.

Author: It seemed most natural to me. First I tried the query setting but that would mean 2 compiled modules OR lazy loading during query time. I didn't like both options much so I just moved to function level where serialization_format is already defined.

Server level - you said it, you cannot differentiate between modules then
Module level - what if we want to have some functions from module without fuel (small, cheap functions) and some with fuel (heavier functions).
Also I assume we would need to extend table schema?

Good comment for access control, I guess we should have a server setting to forbid creating UDFs without fuel if someone wants that?
